### ISSUE TITLE
Calibrate provided pred_probs for classification

### ISSUE THREAD

**Issue by jwmueller** [cleanlab/cleanlab#568]
Can consider the following settings: 

- [ ] standad classification
- [ ] data labeled by multiple annotators
- [ ] multi-label classification

**Comment by turtlemonvh** [comment]
_( Sorry for the long post, but I think this is relevant to the topic of calibration. )_

I was fiddling around with cleanlab on a highly imbalanced binary classification dataset (fraud detection) and ran into issues which *may* be related to calibration.

Taking the simplest approach, I started with this to grab CSVs of predicted label errors and label score probabilities:

```python
# Scores and labels for a temporal hold out set; classifier implementation is all proprietary
df = pd.read_csv("my_scores.csv")

labels = df["Label"]
scores = df[["Score"]].copy()
scores["NF_Score"] = 1 - scores["Score"]

# I ran with both `self_confidence` and `normalized_margin` and the results were ~identical
# https://docs.cleanlab.ai/stable/cleanlab/filter.html#cleanlab.filter.find_label_issues
ordered_label_issues = find_label_issues(labels.to_numpy(), pred_probs=scores.to_numpy(),return_indices_ranked_by='self_confidence', multi_label=False)

pd.Series(ordered_label_issues).to_csv("ordered_label_issues.csv")

label_quality_scores = cleanlab.rank.get_label_quality_scores(labels.to_numpy(), pred_probs=scores.to_numpy(), method='self_confidence', adjust_pred_probs=False)

pd.Series(label_quality_scores).to_csv("label_quality_scores.csv")
```

That CSV contains 330598 lines, 99.4% of the 332612 line data set.

The CDF of those scores looks like this:

![label_quality_scores_cdf_0](https://user-images.githubusercontent.com/889442/236991914-c4f02ec0-9a7a-4906-895b-53b925190198.png)

The CDF is not much better when I set `adjust_pred_probs` to `True`.

![label_quality_scores_adj_cdf_0](https://user-images.githubusercontent.com/889442/236992945-56893b90-f92b-4adf-ab25-02692ebcd4e3.png)

It appears cleanlab thinks something has gone terribly wrong. The data set is 99.8% non-fraud (not uncommon in fraud).

I know that the scores from our typical classifiers are NOT well calibrated to probabilities, so I did try some simple calibration using sklearn's isotonic regression.


```python
ir = IsotonicRegression(out_of_bounds ='clip', increasing=True)
ir.fit(df["Score"], df["Label"])
p_calibrated = ir.transform(df["Score"])
df["Calibrated_Score"] = p_calibrated
```

To convince you the calibration did something, here are before and after plots for that (via `CalibrationDisplay.from_predictions`).

![uncalibrated_scores](https://user-images.githubusercontent.com/889442/236993567-e0d77b0e-382c-4889-a1ab-ee1ad14b2a5b.png)
![calibrated_scores](https://user-images.githubusercontent.com/889442/236993571-3f7fee54-58a8-4c4f-9f8d-03e472d7a891.png)

The istonic regression ended up being very choppy here because of the small number of labels in the hold out set.

I then repeated the analysis, using the new calibrated scores. The csv with the label issues had even more lines, 330731 (vs uncalibrated 330598), but this difference looked like noise. The label quality score CDFs are still quite odd.

![label_quality_scores_cdf_A](https://user-images.githubusercontent.com/889442/236994228-c1aada3d-0f67-438f-ba74-77bf5477c3b4.png)

Zooming in:

![label_quality_scores_cdf_B](https://user-images.githubusercontent.com/889442/236994460-243ac1e0-1f01-42d6-ab89-c31c2188a014.png)

Again, choppy due to the calibration function, but clearly not providing much of an improvement over the uncalibrated score.

## Summary

I heard about cleanlab a while ago, but [a podcast I listened to over the weekend](https://mlops.community/watch/cleanlab--labeled-datasets-that-correct-themselves-automatically_5SHWwmfOQeYtAu/) got me excited about looking more. I was very curious to see what would pop out if all I fed cleanlab was labels and scores.

The results so far have been not great (I'm pretty sure >99% of the data isn't mislabeled 😉 ), and it's not clear to me if this is due to something silly I'm doing or more of a limitation of the approach when applied against heavily imbalanced data.

As a disclaimer: I haven't looked into cleanlab's recommendations in detail. Perhaps the items with the very lowest data quality scores are actually problematic, and if I all got out of cleanlab was an an extension of the "look at the examples where we are most confident, and wrong" approach to data cleaning, but in a less iterative approach, that could still be a win, but when it tells me 99% of my data is suspicious I get suspicious back!

Further disclaimer: I still need to go back and read some of the theory work behind what's going on here, or at least read the code. I may be missing something obvious about how the method works.

**Comment by turtlemonvh** [comment]
**Quick update**: I dug into the code a bit, and I at least get what's going on with `cleanlab.rank.get_label_quality_scores` now. 

Self confidence and normalized margin are [a lot less fancy than I was expecting](https://github.com/cleanlab/cleanlab/blob/v2.3.1/cleanlab/rank.py#L477-L551). Since one is ~just the score and the other is ~just the score - the secondary class score, what I'm seeing above makes a lot of sense.

I think the bigger question is about threshold selection in [find_label_issues](https://github.com/cleanlab/cleanlab/blob/v2.3.1/cleanlab/filter.py#L73) and how to get that to behave for heavily imbalanced data. Doesn't looks like the answer is `filter_by`:

```python
In [106]: filter_by_options = [
     ...:         "low_normalized_margin",
     ...:         "low_self_confidence",
     ...:         "prune_by_noise_rate",
     ...:         "prune_by_class",
     ...:         "both",
     ...:         "confident_learning",
     ...:         "predicted_neq_given",
     ...:     ]

In [107]: for filter_by in filter_by_options:
     ...:     ordered_label_issues = find_label_issues(labels.to_numpy(), pred_probs=scores.to_numpy(), return_indices_ranked_by='normalized_margin', multi_label=False, filter_by=filter_by)
     ...:     print(f"filter_by: {filter_by}\t {len(ordered_label_issues)}")
     ...:
filter_by: low_normalized_margin         330730
filter_by: low_self_confidence   330730
filter_by: prune_by_noise_rate   330730
filter_by: prune_by_class        330730
filter_by: both  330730
filter_by: confident_learning    330730
filter_by: predicted_neq_given   331898
```

Decreasing `frac_noise` (from its default of `1.0`) would definitely get me less samples, but this feels a bit uninteresting since I'm just taking the top `frac_noise * number_of_errors` items.

I'll read more code and some of your papers later. But in the meanwhile, any tips would be helpful!

**Comment by turtlemonvh** [comment]
( I had previously posted a long message here about imbalanced data and getting strange results. This was due to a very silly error - I had the order of the columns backwards in `pred_probs`. )

I have some relevant information from observations while using cleanlab to look at a heavily imbalanced fraud dataset (0.2% positive labels). The data comes from a proprietary model, but we know the labels are mostly right and the score is not calibrated to a probability.

I calibrated the model using isotonic regression - that looks like this:

```python
ir = IsotonicRegression(out_of_bounds ='clip', increasing=True)
ir.fit(df["Score"], df["Label"])
p_calibrated = ir.transform(df["Score"])
df["Calibrated_Score"] = p_calibrated

## Check fit before and after

# Before
disp = CalibrationDisplay.from_predictions(df["Label"], df["Score"], n_bins=20, strategy='uniform')
plt.savefig('uncalibrated_scores.png')

# After
disp = CalibrationDisplay.from_predictions(df["Label"], df["Calibrated_Score"], n_bins=20, strategy='uniform')
plt.savefig('calibrated_scores.png')
```

The changes look like this. They illustrate just how poorly calibrated the model was. This is of course quite common in imbalanced learning scenarios, even when resampling, reweighting, etc.

![uncalibrated_scores](https://user-images.githubusercontent.com/889442/237019332-0bc97b68-5809-4484-920a-2a947204c8d8.png)
![calibrated_scores](https://user-images.githubusercontent.com/889442/237019359-1a3a569f-6d7e-4159-9f42-de76d5736b34.png)

I ran a quick experiment with different values of "filter_by" passed to "find_label_issues" with the calibrated and non-calibrated models to check the difference in behavior.

```python
## Non-calibrated

In [118]: scores = df[["NF_Score", "Score"]].copy()
In [119]: filter_by_options = [
     ...:         "low_normalized_margin",
     ...:         "low_self_confidence",
     ...:         "prune_by_noise_rate",
     ...:         "prune_by_class",
     ...:         "both",
     ...:         "confident_learning",
     ...:         "predicted_neq_given",
     ...:     ]

# Note that `adjust_pred_probs` (and other kwargs) are not passed down to `get_label_quality_scores`, so the `rank_by_kwargs` are only relevant for ranking and not the scoring function used by low_normalized_margin or low_self_confidence
# https://github.com/cleanlab/cleanlab/blob/v2.3.1/cleanlab/filter.py#L346-L352

In [120]: for filter_by in filter_by_options:
     ...:     ordered_label_issues = find_label_issues(labels.to_numpy(), pred_probs=scores.to_numpy(), return_indices_ranked_by='normalized_margin', multi_label=False, filter_by=filter_by, rank_by_kwargs={"adjust_pred_probs": False})
     ...:     print(f"filter_by: {filter_by}\t {len(ordered_label_issues)}")
     ...:
filter_by: low_normalized_margin         1033
filter_by: low_self_confidence   1033
filter_by: prune_by_noise_rate   1224
filter_by: prune_by_class        1224
filter_by: both  1224
filter_by: confident_learning    1033
filter_by: predicted_neq_given   2014

## Try calibrated data

In [127]: df["NF_Calibrated_Score"] = 1.0 - df["Calibrated_Score"]

In [128]: scores = df[["NF_Calibrated_Score", "Calibrated_Score"]].copy()
In [129]: for filter_by in filter_by_options:
     ...:     ordered_label_issues = find_label_issues(labels.to_numpy(), pred_probs=scores.to_numpy(), return_indices_ranked_by='normalized_margin', multi_label=False, filter_by=filter_by, rank_by_kwargs={"adjust_pred_probs": False})
     ...:     print(f"filter_by: {filter_by}\t {len(ordered_label_issues)}")
     ...:
filter_by: low_normalized_margin         68
filter_by: low_self_confidence   68
filter_by: prune_by_noise_rate   75
filter_by: prune_by_class        75
filter_by: both  75
filter_by: confident_learning    68
filter_by: predicted_neq_given   713
```

I haven't looked at the data alongside the recommendations from cleanlab, but I suspect the second (calibrated) set of answers is closer to being correct, and the fact that the size of the recommendations is >10x different when using the calibrated model is certainly interesting. If the results are "more correct" using calibration, including that would certainly make using this package easier for others like me in the future.

**Comment by cgnorthcutt** [comment]
Very interesting @turtlemonvh thanks for sharing! Do you know a priori how much error is in the dataset? Also would you be able to randomly sample 30 of the errors from the calibrated and non calibrated results and report the accuracy of whether it actually is an error for each to help better inform what we do next?

**Comment by turtlemonvh** [comment]
@cgnorthcutt Unfortunately this data concerns potential fraud which happened several months ago; getting a corrected final disposition for those transactions would require work from a partner organization to dig into suspicious cases, which would be prohibitively expensive. Anecdotally I expect the label error rate to be low, with FN >> FP. ~70 errors in a dataset with ~700 observed fraud does make sense here. ~1000 errors is definitely too high.

There are some data sets we'll be working on soon where we will be able to make stronger assumptions about the fraction of missing labels (e.g., bc the label generating process is necessarily lossy for historical payment system reasons) - we'll look for opportunities to use cleanlab to reduce bias from these FNs. We also sometimes have cases where labeling is a fuzzy join to the original dataset because the label generating process doesn't provide enough details to uniquely identify a transaction - cleanlab could help us avoid the much more problematic FPs in our training set there.

But back to modeling the effect: I could introduce some known noise into the data via label flipping and see how well cleanlab identifies these as errors, but I imagine this is something you can do on your end already.

**Comment by cgnorthcutt** [comment]
Thanks for the explanation. I'd recommend you at least try [Cleanlab Studio](Cleanlab.ai). It will visualize the mistakes and provide analysis of the false positives. Should help you hone down on some of the discrepancies you described.
